### PR TITLE
Dashboard widget: Activity by others

### PR DESCRIPTION
**Pull request by jakobroehrl**
Signed-off-by: Jakob Röhrl <jakob.roehrl@web.de>
fixes: #579

image

### CODE REVIEW

**Comment by artonge** [comment]
Hey @jakobroehrl , after a quick glance at your code, you are not registering the dashboard widget. From the PR you linked: https://github.com/nextcloud/tasks/pull/1641/files#diff-4eb824426a8cffb8b33b62b57744c69f32ea30ea755249a7a6170d9402885debR44

File: src/dashboard.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ */
+
+import Dashboard from './views/Dashboard'
+// import store from './store/store.js'
+
+// import './css/dashboard.scss'
+
+import { generateFilePath } from '@nextcloud/router'
+import { getRequestToken } from '@nextcloud/auth'
```
Comment (nickvergessen) [inline src/dashboard.js:29]: "@nextcloud/auth" is extraneous

so add to package.json

File: src/views/Dashboard.vue [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		},
+
+		processActivities({ data }) {
+			if (data.ocs && data.ocs.data && data.ocs.data.length > 0) {
+				// create Activity objects and sort by newest
+				this.activities = data.ocs.data
+					.map(activity => new ActivityModel(activity))
+					.sort((a, b) => b.timestamp - a.timestamp)
+			}
+			console.log(data)
```
Comment (nickvergessen) [inline src/views/Dashboard.vue:96]:
```suggestion
```

File: src/views/Dashboard.vue [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+				console.error('Error loading the activity list', error)
+			}
+		},
+
+		processActivities({ data }) {
+			if (data.ocs && data.ocs.data && data.ocs.data.length > 0) {
+				// create Activity objects and sort by newest
+				this.activities = data.ocs.data
+					.map(activity => new ActivityModel(activity))
+					.sort((a, b) => b.timestamp - a.timestamp)
```
Comment (nickvergessen) [inline src/views/Dashboard.vue:94]: the data is sorted already. Maybe use reverse instead.

File: src/views/Dashboard.vue [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		:half-empty-content-message="t('activity', 'No activities')">
+		<template #default="{ item }">
+			<!-- {{item._activity}} -->
+			<DashboardWidgetItem
+				:id="item.activity_id"
+				:item="item._activity"
+				:target-url="item.link"
+				:avatar-username="item.user"
+				:main-text="item.subject"
+				:sub-text="item.datetime" />
```
Comment (nickvergessen) [inline src/views/Dashboard.vue:40]: this needs translating to a decent time vision?
Bildschirmfoto von 2021-07-01 11-49-35

File: lib/AppInfo/Application.php [hunk 5]
```diff
@@ -40,8 +41,8 @@
 class Application extends App implements IBootstrap {
 	public const APP_ID = 'activity';
 
-	public function __construct() {
-		parent::__construct(self::APP_ID);
+	public function __construct(array $params=[]) {
```
Comment (nickvergessen) [inline lib/AppInfo/Application.php:44]:
```suggestion
	public function __construct(array $params = []) {
```

File: src/views/Dashboard.vue [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+					this.loading = false
+					return
+				}
+				this.loading = false
+				console.error('Error loading the activity list', error)
+			}
+		},
+
+		processActivities({ data }) {
+			if (data.ocs && data.ocs.data && data.ocs.data.length > 0) {
```
Comment (nickvergessen) [inline src/views/Dashboard.vue:109]: Missing else branch to clear it?

File: src/views/Dashboard.vue [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		:half-empty-content-message="t('activity', 'No activities')">
+		<template #default="{ item }">
+			<!-- {{item._activity}} -->
+			<DashboardWidgetItem
+				:id="item.activity_id"
+				:target-url="item.link"
+				:avatar-username="item.user"
+				:overlay-icon-url="item.icon"
+				:main-text="item.dateFromNow"
+				:sub-text="item.subject" />
```
Comment (nickvergessen) [inline src/views/Dashboard.vue:40]:
```suggestion
				:main-text="item.subject"
				:sub-text="item.dateFromNow" />
```

The action is more important than the when.

File: lib/Dashboard/ActivityWidget.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	public function getId(): string {
+		return Application::APP_ID;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getTitle(): string {
+		return $this->l10n->t('Latest activities');
```
Comment (nickvergessen) [inline lib/Dashboard/ActivityWidget.php:58]: I would go with 

```suggestion
		return $this->l10n->t('Recent activity');
```

to be in with other widgets

File: src/views/Dashboard.vue (hunk 4 above)
Comment (jakobroehrl) [reply src/views/Dashboard.vue]: @nickvergessen Thanks, changed all your points!

**Comment by nickvergessen** [comment]
Once done you need to submit `make build-js-production`

**Comment by jakobroehrl** [comment]
> Once done you need to submit `make build-js-production`

done!

**Comment by nickvergessen** [comment]
Needs special handling for the activity-icons on dark mode

Helping selector if you do it in the activity app:
```css
body.theme--dark &
```

Server ref is https://github.com/nextcloud/server/blob/f83de9d1943af116e89882aa577bd75c87b0c966/apps/accessibility/css/dark.scss#L27-L33

**Comment by jakobroehrl** [comment]
@nickvergessen 
Thanks!
How could I only show actions from others? Maybe this would be even better?

**Comment by jakobroehrl** [comment]
> @nickvergessen
> Thanks!
> How could I only show actions from others? Maybe this would be even better?

Figured out myself. Solved it with a simple filter.
Should I create a all activities widget or a activities from other or should I create 2 widgets?

**Comment by jancborchardt** [comment]
Looks nice! Some design feedback:
- If you yourself changed something, should the icon really be your avatar? For single-user instances this would result in the same avatar for every entry. Possibly better would be that for actions you did yourself, the app icon shows as the action icon.
- The main line of the text ellipsizes although it is the important info, and since the info is chronologically sorted, the time is sort of redundant in this small space. Here’s a proposal:

> Important event
> edited by [name] · [Calendar name]
> 
> Organizing conference
> edited by [name] · Marketing board
>
> New folder
> shared by [name]

**Comment by nickvergessen** [comment]
> Here’s a proposal:

It's not suitable to break the API or require a new API to be implemented :P

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
@artonge LGTM, want to give it a quick test as well?

File: src/dashboard.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import Vue from 'vue'
+// import Vuex from 'vuex'
+
+// Vue.use(Vuex)
+
+// eslint-disable-next-line
+__webpack_nonce__ = btoa(getRequestToken())
+
+// eslint-disable-next-line
+__webpack_public_path__ = generateFilePath('activity', '', 'js/')
```
Comment (artonge) [inline src/dashboard.js:41]:
```suggestion
__webpack_public_path__ = generateFilePath(appName, '', 'js/')
```

File: src/dashboard.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+Vue.prototype.n = translatePlural
+Vue.prototype.OC = OC
+Vue.prototype.OCA = OCA
+
+document.addEventListener('DOMContentLoaded', () => {
+	OCA.Dashboard.register('activity', (el) => {
+		const View = Vue.extend(Dashboard)
+		const vm = new View({
+			propsData: {},
+			// store,
```
Comment (artonge) [inline src/dashboard.js:53]:
```suggestion
```

**Review by artonge — CHANGES_REQUESTED** [review]
Minor nitpicks

File: src/dashboard.js (hunk 9 above)
Comment (jakobroehrl) [reply src/dashboard.js:41]: sorry, this is not working.

**Comment by jakobroehrl** [comment]
Let's merge it in. :)

File: src/dashboard.js (hunk 9 above)
Comment (artonge) [reply src/dashboard.js:41]: What do you mean by not working ? Just tried and it works fine :)

Comment (jakobroehrl) [reply src/dashboard.js]: Now it's working. Tried 'appName' instead of appName.

**Review by artonge — APPROVED** [review]

**Pull request merged**